integration-tests/jbang/src/test/java/io/quarkus/jbang/RunJBangTest.java

```diff
@@ -0,0 +1,155 @@
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+import org.zeroturnaround.exec.ProcessExecutor;
+import org.zeroturnaround.exec.StartedProcess;
+import org.zeroturnaround.exec.stream.LogOutputStream;
+
+class RunJBangTest {
+
+    private static final String LATEST_DOWNLOAD_JBANG_ZIP = "https://www.jbang.dev/releases/latest/download/jbang.zip";
+    private static Path DEVMODE_SCRIPT;
+    private static Path JBANG_DIR = null;
+
+    /**
+     * Downloads latest jbang in static dir and setup JBANG_DIR to be 100% isolated
+     * from whatever jbang could otherwise be installed on the system.
+     */
+    @BeforeAll
+    static void downloadLatestJBang(@TempDir Path tempdir) throws IOException {
+        InputStream in = new URL(LATEST_DOWNLOAD_JBANG_ZIP).openStream();
+        JBANG_DIR = tempdir.resolve(".jbang");
+        Path zipfile = tempdir.resolve("jbang.zip");
+        Files.copy(in, zipfile, StandardCopyOption.REPLACE_EXISTING);
+        TestHelper.unzip(zipfile, JBANG_DIR, true, null);
+
+        //copy over files to allow edit
+        DEVMODE_SCRIPT = tempdir.resolve("devmode.java");
+        Files.copy(new File("target/test-classes/devmode.java").toPath(), DEVMODE_SCRIPT, StandardCopyOption.REPLACE_EXISTING);
+    }
+
+    private ProcessExecutor getJBangExecutor(String... args) {
+        String cmd;
+        if (isWindows()) {
+            cmd = "bin/jbang.cmd";
+        } else {
+            cmd = "bin/jbang";
+        }
+
+        List<String> realargs = new ArrayList<>();
+        realargs.add(JBANG_DIR.resolve(cmd).toString());
+        realargs.addAll(Arrays.asList(args));
+
+        return new ProcessExecutor().command(realargs)
+                .environment("JBANG_DIR", JBANG_DIR.toAbsolutePath().toString())
+                .redirectErrorStream(true).readOutput(true);
+    }
+
+    /**
+     * just a sanity check that jbang is behaving as we expect with respect to config
+     **/
+    @Test
+    void testRunJBangWithProperDir() throws IOException, InterruptedException, TimeoutException {
+
+        String result = getJBangExecutor("version", "--verbose")
+                .exitValue(0)
+                .execute().outputUTF8();
+        assertThat(result).contains(JBANG_DIR.toAbsolutePath().toString());
+    }
+
+    @Test
+    void testRunJBangInDevMode() throws Throwable {
+
+        final CompletableFuture<Integer> listening = new CompletableFuture<>();
+        final CountDownLatch devmode = new CountDownLatch(1);
+        final CountDownLatch replaced = new CountDownLatch(1);
+
+        StartedProcess exec = getJBangExecutor("-Dquarkus.dev", "-Dq.v=999-SNAPSHOT",
+                "-Dquarkus.http.port=0",
```

Comment:
I don't know whether this is still a problem but @geoand had to roll back something like that not too long ago: https://github.com/quarkusio/quarkus/commit/f80a389de25572d2c29dd5af1ec567acfe46b1fb

Reply:
@maxandersen why is this using a random port?

Reply:
IIRC, I disabled the random port because it was causing problems with CI.

I think I also did something similar recently with DevServices

Reply:
Avoid conflicts between restarts if kills doesn't work. 

I had initially a random port from 8000-9000.

But surprised :0 should ever give conflict as it's the OS assigning its free port ?

Reply:
That was why I changed it to 0, as I figured it would be more reliable than the 8000-9000 approach.

Reply:
Not sure this applies to Quarkus CI, but I had such problems when the you (or a service) is using fixed ports that lie in the range of ports the operating system is using to assign random ports.
See also `proc/sys/net/ipv4/ip_local_port_range`.

Reply:
FWIW, ubuntu-latest, runner version: '2.278.0':
```
cat /proc/sys/net/ipv4/ip_local_port_range
32768	60999
```

Reply:
im not sure what the ask is here ? 

want me to bring back a random number from a smaller range or ?

Reply:
I would leave it like this and see how it goes. If it fails I can try to invest some time to find out what's going on... to fix it for good!